include/seqan3/io/structure_file/detail.hpp

```diff
@@ -0,0 +1,109 @@
+// ============================================================================
+//                 SeqAn - The Library for Sequence Analysis
+// ============================================================================
+//
+// Copyright (c) 2006-2018, Knut Reinert & Freie Universitaet Berlin
+// Copyright (c) 2016-2018, Knut Reinert & MPI Molekulare Genetik
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//     * Neither the name of Knut Reinert or the FU Berlin nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+/*!\file
+ * \brief Helper functions (e.g. conversions) for the structure IO submodule.
+ * \author Jörg Winkler <j.winkler AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <map>
+#include <stack>
+
+#include <seqan3/alphabet/structure/rna_structure_concept.hpp>
+
+namespace seqan3::detail
+{
+/*!
+ * \brief Transforms a structure annotation string into a base pair probability matrix.
+ * \throws seqan3::parse_error if unpaired brackets are found in the structure annotation.
+ * \tparam structure_alph_type The type of the structure alphabet; must satisfy seqan3::rna_structure_concept.
+ * \tparam bpp_type            The type of the target matrix.
+ * \tparam structure_type      The range type of the structure annotation.
+ * \param[out] bpp             The target matrix that receives the base pair probabilities.
```

Comment:
Why don't you return this instead of doing an out-parameter?

Reply:
Ah, because you need to specify the type?

Reply:
Good idea, I give the type by template argument and return `bpp`.

Reply:
Ah, no, sorry, I meant that if you need to able specify the type, than it's perfectly fine to use an out-parameter and likely a better solution that having to pass an extra template parameters.

Reply:
Could you revert this change? Otherwise it looks good now.